server.js

```diff
@@ -91,6 +91,7 @@ if (config.blockedAtWarnEnable) {
 const app = express();
 app.set('views', path.join(__dirname, 'pages'));
 app.set('view engine', 'ejs');
+app.set('trust proxy', 'loopback');
```

Comment:
Doesn't this have the same issue as the same proposed change to #1697? See https://github.com/PrairieLearn/PrairieLearn/pull/1697#discussion_r323416628

Reply:
My preference was to piggy back on existing tools that do this kind of manipulation rather than doing it ourselves, that's why I brought it back. I took a few minutes today to dig under the layers and see what it's doing.

`express` calls `proxy-addr` to do the parsing, which calls `forwarded` (these are both part of https://jshttp.github.io/) Digging back through the code gave me the ability to setup a quick test to see what happens if we run the `trust proxy` line above with a spoofed x-forwarded-for header, both in direct connection mode (like with docker) and when running behind the apache proxy.

```
Direct with spoofed x-forwarded-for
req { headers: { 'x-forwarded-for': '9.9.9.9' },
  connection: { remoteAddress: '128.174.55.55' } }
forwarded(req) [ '128.174.55.55', '9.9.9.9' ]
proxyaddr(req, 'loopback') 128.174.55.55


Proxy with spoofed x-forwarded-for trusting loopback
req { headers: { 'x-forwarded-for': '9.9.9.9, 128.174.55.55' },
  connection: { remoteAddress: '::1' } }
forwarded(req) [ '::1', '128.174.55.55', '9.9.9.9' ]
proxyaddr(req, 'loopback') 128.174.55.55
```

The proxy case mimics the headers received by PL under `pl-dev`. In both cases, it came out with the correct (non-spoofed) proxyaddr address. Showing the forwarded step probably isn't necessary, but it shows how the order of "forwarded" IPs is flipped between the HTTP header (appends to the end of the header string when new forwards are added) vs the JS array (which puts the most-recently seen IP at index 0).

Because apache interjects the connected IP before anything else it received in the HTTP header, it's "closest" to localhost. The 'loopback' argument tells proxyaddr to shift from the front of the array the first entry that's not one of the localhost addresses.

So I think we're good?

Reply:
Cool, thanks for digging into the details here!

I agree that this is safe, given that we are only trusting `loopback` as a proxy.